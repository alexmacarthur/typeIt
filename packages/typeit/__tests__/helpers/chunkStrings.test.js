import {
  maybeChunkStringAsHtml,
  walkElementNodes,
} from "../../src/helpers/chunkStrings";
import GraphemeSplitter from "grapheme-splitter";
import {
  defaultChunkStringAsHtml,
  defaultExpandTextNodes,
  defaultMaybeChunkStringAsHtml,
} from "./util";

beforeEach(() => {
  document.body.innerHTML = "";
});

test("Parses normal string correctly.", () => {
  let result = defaultChunkStringAsHtml("Hello, this is my string.");

  expect(result).toMatchSnapshot();
});

test("Parses single HTML tag.", () => {
  let result = defaultChunkStringAsHtml(
    "Hello, this is some <strong>bold</strong> text."
  );

  expect(result).toMatchSnapshot();
});

test("Parses multiple HTML tags.", () => {
  let result = defaultChunkStringAsHtml(
    "Hello, this is some <strong>bold</strong> text, and some <i>italicized</i> text."
  );

  expect(result).toMatchSnapshot();
});

test("Parses HTML tag at beginning of string.", () => {
  let result = defaultChunkStringAsHtml(
    "<strong>Hello!</strong> This is some text with HTML at the beginning."
  );

  expect(result).toMatchSnapshot();
});

test("Parses HTML tag at end of string.", () => {
  let result = defaultChunkStringAsHtml(
    "This is some text with HTML at the <em>end.</em>"
  );

  expect(result).toMatchSnapshot();
});

test("Parses HTML tag with attributes.", () => {
  let result = defaultChunkStringAsHtml(
    'This string has an <strong class="strong-class" id="strong-id" data-whatever="data-att">element</strong> with attributes.'
  );

  expect(result).toMatchSnapshot();
});

describe("maybeChunkStringAsHtml()", () => {
  test("Should return noderized string when setting is enabled.", () => {
    expect(
      defaultMaybeChunkStringAsHtml("A <em>fancy</em> string.")
    ).toMatchSnapshot();
  });

  test("Should correctly transform non-HTML string as character objects.", () => {
    expect(
      defaultMaybeChunkStringAsHtml("A <em>fancy</em> string.", false)
    ).toMatchSnapshot();
  });
  test("Should return correctly split string with emoji.", () => {
    const splitter = new GraphemeSplitter();
    expect(
      maybeChunkStringAsHtml(
        "An 🍕 emoji 🌷 filled 👍 string. ⬆️",
        false,
        (str) => splitter.splitGraphemes(str)
      )
    ).toMatchSnapshot();
  });
});

describe("walkElementNodes()", () => {
  test("Gathers correct number of nodes with original parents attached.", () => {
    setHTML`<h2 id="heading">A<span>B<small>C</small></span></h2>`;

    const el = defaultExpandTextNodes(document.getElementById("heading"));
    const result = walkElementNodes(el);
    const thirdNode = result[2];

    // "B" should have an original parent of the <span>.
    expect(thirdNode.originalParent).toMatchObject(
      document.querySelector("span")
    );

    // Includes each text node, and separate nodes for the parent elements themselves.
    expect(result.length).toEqual(5);
  });

  test("Reverses nodes when parameter is passed.", () => {
    setHTML`<h2 id="heading">ABC</h2>`;

    const el = defaultExpandTextNodes(document.getElementById("heading"));
    const result = walkElementNodes(el, true);

    expect(result[0].textContent).toEqual("C");
  });

  test("Excludes cursor.", () => {
    setHTML`<h2 id="heading"><span>123</span><i class="ti-cursor">|</i></h2>`;

    const el = defaultExpandTextNodes(document.getElementById("heading"));
    const result = walkElementNodes(el);
    const cursor = result.find((n) => n?.classList?.contains("ti-cursor"));

    expect(cursor).toBe(undefined);
    expect(result).toHaveLength(4);
  });
});
