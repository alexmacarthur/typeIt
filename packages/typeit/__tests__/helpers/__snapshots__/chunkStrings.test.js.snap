// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parses HTML tag at beginning of string. 1`] = `
Array [
  Object {
    "content": "H",
    "node": Hello!,
  },
  Object {
    "content": "e",
    "node": Hello!,
  },
  Object {
    "content": "l",
    "node": Hello!,
  },
  Object {
    "content": "l",
    "node": Hello!,
  },
  Object {
    "content": "o",
    "node": Hello!,
  },
  Object {
    "content": "!",
    "node": Hello!,
  },
  Object {
    "content": " ",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "T",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "h",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "i",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "s",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": " ",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "i",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "s",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": " ",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "s",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "o",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "m",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "e",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": " ",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "t",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "e",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "x",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "t",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": " ",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "w",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "i",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "t",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "h",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": " ",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "H",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "T",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "M",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "L",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": " ",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "a",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "t",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": " ",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "t",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "h",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "e",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": " ",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "b",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "e",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "g",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "i",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "n",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "n",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "i",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "n",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": "g",
    "node":  This is some text with HTML at the beginning.,
  },
  Object {
    "content": ".",
    "node":  This is some text with HTML at the beginning.,
  },
]
`;

exports[`Parses HTML tag at end of string. 1`] = `
Array [
  Object {
    "content": "T",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "h",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "i",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "s",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": " ",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "i",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "s",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": " ",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "s",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "o",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "m",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "e",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": " ",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "t",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "e",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "x",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "t",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": " ",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "w",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "i",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "t",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "h",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": " ",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "H",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "T",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "M",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "L",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": " ",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "a",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "t",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": " ",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "t",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "h",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "e",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": " ",
    "node": This is some text with HTML at the ,
  },
  Object {
    "content": "b",
    "node": beginning.,
  },
  Object {
    "content": "e",
    "node": beginning.,
  },
  Object {
    "content": "g",
    "node": beginning.,
  },
  Object {
    "content": "i",
    "node": beginning.,
  },
  Object {
    "content": "n",
    "node": beginning.,
  },
  Object {
    "content": "n",
    "node": beginning.,
  },
  Object {
    "content": "i",
    "node": beginning.,
  },
  Object {
    "content": "n",
    "node": beginning.,
  },
  Object {
    "content": "g",
    "node": beginning.,
  },
  Object {
    "content": ".",
    "node": beginning.,
  },
]
`;

exports[`Parses HTML tag with attributes. 1`] = `
Array [
  Object {
    "content": "T",
    "node": This string has an ,
  },
  Object {
    "content": "h",
    "node": This string has an ,
  },
  Object {
    "content": "i",
    "node": This string has an ,
  },
  Object {
    "content": "s",
    "node": This string has an ,
  },
  Object {
    "content": " ",
    "node": This string has an ,
  },
  Object {
    "content": "s",
    "node": This string has an ,
  },
  Object {
    "content": "t",
    "node": This string has an ,
  },
  Object {
    "content": "r",
    "node": This string has an ,
  },
  Object {
    "content": "i",
    "node": This string has an ,
  },
  Object {
    "content": "n",
    "node": This string has an ,
  },
  Object {
    "content": "g",
    "node": This string has an ,
  },
  Object {
    "content": " ",
    "node": This string has an ,
  },
  Object {
    "content": "h",
    "node": This string has an ,
  },
  Object {
    "content": "a",
    "node": This string has an ,
  },
  Object {
    "content": "s",
    "node": This string has an ,
  },
  Object {
    "content": " ",
    "node": This string has an ,
  },
  Object {
    "content": "a",
    "node": This string has an ,
  },
  Object {
    "content": "n",
    "node": This string has an ,
  },
  Object {
    "content": " ",
    "node": This string has an ,
  },
  Object {
    "content": "e",
    "node": element,
  },
  Object {
    "content": "l",
    "node": element,
  },
  Object {
    "content": "e",
    "node": element,
  },
  Object {
    "content": "m",
    "node": element,
  },
  Object {
    "content": "e",
    "node": element,
  },
  Object {
    "content": "n",
    "node": element,
  },
  Object {
    "content": "t",
    "node": element,
  },
  Object {
    "content": " ",
    "node":  with attributes.,
  },
  Object {
    "content": "w",
    "node":  with attributes.,
  },
  Object {
    "content": "i",
    "node":  with attributes.,
  },
  Object {
    "content": "t",
    "node":  with attributes.,
  },
  Object {
    "content": "h",
    "node":  with attributes.,
  },
  Object {
    "content": " ",
    "node":  with attributes.,
  },
  Object {
    "content": "a",
    "node":  with attributes.,
  },
  Object {
    "content": "t",
    "node":  with attributes.,
  },
  Object {
    "content": "t",
    "node":  with attributes.,
  },
  Object {
    "content": "r",
    "node":  with attributes.,
  },
  Object {
    "content": "i",
    "node":  with attributes.,
  },
  Object {
    "content": "b",
    "node":  with attributes.,
  },
  Object {
    "content": "u",
    "node":  with attributes.,
  },
  Object {
    "content": "t",
    "node":  with attributes.,
  },
  Object {
    "content": "e",
    "node":  with attributes.,
  },
  Object {
    "content": "s",
    "node":  with attributes.,
  },
  Object {
    "content": ".",
    "node":  with attributes.,
  },
]
`;

exports[`Parses multiple HTML tags. 1`] = `
Array [
  Object {
    "content": "H",
    "node": Hello, this is some ,
  },
  Object {
    "content": "e",
    "node": Hello, this is some ,
  },
  Object {
    "content": "l",
    "node": Hello, this is some ,
  },
  Object {
    "content": "l",
    "node": Hello, this is some ,
  },
  Object {
    "content": "o",
    "node": Hello, this is some ,
  },
  Object {
    "content": ",",
    "node": Hello, this is some ,
  },
  Object {
    "content": " ",
    "node": Hello, this is some ,
  },
  Object {
    "content": "t",
    "node": Hello, this is some ,
  },
  Object {
    "content": "h",
    "node": Hello, this is some ,
  },
  Object {
    "content": "i",
    "node": Hello, this is some ,
  },
  Object {
    "content": "s",
    "node": Hello, this is some ,
  },
  Object {
    "content": " ",
    "node": Hello, this is some ,
  },
  Object {
    "content": "i",
    "node": Hello, this is some ,
  },
  Object {
    "content": "s",
    "node": Hello, this is some ,
  },
  Object {
    "content": " ",
    "node": Hello, this is some ,
  },
  Object {
    "content": "s",
    "node": Hello, this is some ,
  },
  Object {
    "content": "o",
    "node": Hello, this is some ,
  },
  Object {
    "content": "m",
    "node": Hello, this is some ,
  },
  Object {
    "content": "e",
    "node": Hello, this is some ,
  },
  Object {
    "content": " ",
    "node": Hello, this is some ,
  },
  Object {
    "content": "b",
    "node": bold,
  },
  Object {
    "content": "o",
    "node": bold,
  },
  Object {
    "content": "l",
    "node": bold,
  },
  Object {
    "content": "d",
    "node": bold,
  },
  Object {
    "content": " ",
    "node":  text, and some ,
  },
  Object {
    "content": "t",
    "node":  text, and some ,
  },
  Object {
    "content": "e",
    "node":  text, and some ,
  },
  Object {
    "content": "x",
    "node":  text, and some ,
  },
  Object {
    "content": "t",
    "node":  text, and some ,
  },
  Object {
    "content": ",",
    "node":  text, and some ,
  },
  Object {
    "content": " ",
    "node":  text, and some ,
  },
  Object {
    "content": "a",
    "node":  text, and some ,
  },
  Object {
    "content": "n",
    "node":  text, and some ,
  },
  Object {
    "content": "d",
    "node":  text, and some ,
  },
  Object {
    "content": " ",
    "node":  text, and some ,
  },
  Object {
    "content": "s",
    "node":  text, and some ,
  },
  Object {
    "content": "o",
    "node":  text, and some ,
  },
  Object {
    "content": "m",
    "node":  text, and some ,
  },
  Object {
    "content": "e",
    "node":  text, and some ,
  },
  Object {
    "content": " ",
    "node":  text, and some ,
  },
  Object {
    "content": "i",
    "node": italicized,
  },
  Object {
    "content": "t",
    "node": italicized,
  },
  Object {
    "content": "a",
    "node": italicized,
  },
  Object {
    "content": "l",
    "node": italicized,
  },
  Object {
    "content": "i",
    "node": italicized,
  },
  Object {
    "content": "c",
    "node": italicized,
  },
  Object {
    "content": "i",
    "node": italicized,
  },
  Object {
    "content": "z",
    "node": italicized,
  },
  Object {
    "content": "e",
    "node": italicized,
  },
  Object {
    "content": "d",
    "node": italicized,
  },
  Object {
    "content": " ",
    "node":  text.,
  },
  Object {
    "content": "t",
    "node":  text.,
  },
  Object {
    "content": "e",
    "node":  text.,
  },
  Object {
    "content": "x",
    "node":  text.,
  },
  Object {
    "content": "t",
    "node":  text.,
  },
  Object {
    "content": ".",
    "node":  text.,
  },
]
`;

exports[`Parses normal string correctly. 1`] = `
Array [
  Object {
    "content": "H",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "e",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "l",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "l",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "o",
    "node": Hello, this is my string.,
  },
  Object {
    "content": ",",
    "node": Hello, this is my string.,
  },
  Object {
    "content": " ",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "t",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "h",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "i",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "s",
    "node": Hello, this is my string.,
  },
  Object {
    "content": " ",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "i",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "s",
    "node": Hello, this is my string.,
  },
  Object {
    "content": " ",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "m",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "y",
    "node": Hello, this is my string.,
  },
  Object {
    "content": " ",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "s",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "t",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "r",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "i",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "n",
    "node": Hello, this is my string.,
  },
  Object {
    "content": "g",
    "node": Hello, this is my string.,
  },
  Object {
    "content": ".",
    "node": Hello, this is my string.,
  },
]
`;

exports[`Parses single HTML tag. 1`] = `
Array [
  Object {
    "content": "H",
    "node": Hello, this is some ,
  },
  Object {
    "content": "e",
    "node": Hello, this is some ,
  },
  Object {
    "content": "l",
    "node": Hello, this is some ,
  },
  Object {
    "content": "l",
    "node": Hello, this is some ,
  },
  Object {
    "content": "o",
    "node": Hello, this is some ,
  },
  Object {
    "content": ",",
    "node": Hello, this is some ,
  },
  Object {
    "content": " ",
    "node": Hello, this is some ,
  },
  Object {
    "content": "t",
    "node": Hello, this is some ,
  },
  Object {
    "content": "h",
    "node": Hello, this is some ,
  },
  Object {
    "content": "i",
    "node": Hello, this is some ,
  },
  Object {
    "content": "s",
    "node": Hello, this is some ,
  },
  Object {
    "content": " ",
    "node": Hello, this is some ,
  },
  Object {
    "content": "i",
    "node": Hello, this is some ,
  },
  Object {
    "content": "s",
    "node": Hello, this is some ,
  },
  Object {
    "content": " ",
    "node": Hello, this is some ,
  },
  Object {
    "content": "s",
    "node": Hello, this is some ,
  },
  Object {
    "content": "o",
    "node": Hello, this is some ,
  },
  Object {
    "content": "m",
    "node": Hello, this is some ,
  },
  Object {
    "content": "e",
    "node": Hello, this is some ,
  },
  Object {
    "content": " ",
    "node": Hello, this is some ,
  },
  Object {
    "content": "b",
    "node": bold,
  },
  Object {
    "content": "o",
    "node": bold,
  },
  Object {
    "content": "l",
    "node": bold,
  },
  Object {
    "content": "d",
    "node": bold,
  },
  Object {
    "content": " ",
    "node":  text.,
  },
  Object {
    "content": "t",
    "node":  text.,
  },
  Object {
    "content": "e",
    "node":  text.,
  },
  Object {
    "content": "x",
    "node":  text.,
  },
  Object {
    "content": "t",
    "node":  text.,
  },
  Object {
    "content": ".",
    "node":  text.,
  },
]
`;

exports[`maybeChunkStringAsHtml() Should correctly transform non-HTML string as character objects. 1`] = `
Array [
  Object {
    "content": "A",
    "node": null,
  },
  Object {
    "content": " ",
    "node": null,
  },
  Object {
    "content": "<",
    "node": null,
  },
  Object {
    "content": "e",
    "node": null,
  },
  Object {
    "content": "m",
    "node": null,
  },
  Object {
    "content": ">",
    "node": null,
  },
  Object {
    "content": "f",
    "node": null,
  },
  Object {
    "content": "a",
    "node": null,
  },
  Object {
    "content": "n",
    "node": null,
  },
  Object {
    "content": "c",
    "node": null,
  },
  Object {
    "content": "y",
    "node": null,
  },
  Object {
    "content": "<",
    "node": null,
  },
  Object {
    "content": "/",
    "node": null,
  },
  Object {
    "content": "e",
    "node": null,
  },
  Object {
    "content": "m",
    "node": null,
  },
  Object {
    "content": ">",
    "node": null,
  },
  Object {
    "content": " ",
    "node": null,
  },
  Object {
    "content": "s",
    "node": null,
  },
  Object {
    "content": "t",
    "node": null,
  },
  Object {
    "content": "r",
    "node": null,
  },
  Object {
    "content": "i",
    "node": null,
  },
  Object {
    "content": "n",
    "node": null,
  },
  Object {
    "content": "g",
    "node": null,
  },
  Object {
    "content": ".",
    "node": null,
  },
]
`;

exports[`maybeChunkStringAsHtml() Should return noderized string when setting is enabled. 1`] = `
Array [
  Object {
    "content": "A",
    "node": A ,
  },
  Object {
    "content": " ",
    "node": A ,
  },
  Object {
    "content": "f",
    "node": fancy,
  },
  Object {
    "content": "a",
    "node": fancy,
  },
  Object {
    "content": "n",
    "node": fancy,
  },
  Object {
    "content": "c",
    "node": fancy,
  },
  Object {
    "content": "y",
    "node": fancy,
  },
  Object {
    "content": " ",
    "node":  string.,
  },
  Object {
    "content": "s",
    "node":  string.,
  },
  Object {
    "content": "t",
    "node":  string.,
  },
  Object {
    "content": "r",
    "node":  string.,
  },
  Object {
    "content": "i",
    "node":  string.,
  },
  Object {
    "content": "n",
    "node":  string.,
  },
  Object {
    "content": "g",
    "node":  string.,
  },
  Object {
    "content": ".",
    "node":  string.,
  },
]
`;
